### PR TITLE
Set network and memory policy on Request.

### PR DESCRIPTION
**Pull request by NightlyNexus**
Not sure how to test this.
https://github.com/square/picasso/issues/1800

Closes https://github.com/square/picasso/issues/1797 but it would be nice to have a test.

note that this lets you set a MemoryPolicy.NO_CACHE in a RequestTransformer which sometimes has an effect and sometimes doesn't.

### CODE REVIEW

**Comment by jrodbx** [comment]
The more I think about this, the more I can't come up with a practical use case.  Maybe we leave this open or close/revisiting later when we have one?

**Comment by NightlyNexus** [comment]
I think you're right. I'll take another look later tonight.

**Comment by NightlyNexus** [comment]
Removed the Request.Builder APIs. Opinions on this as a fix for #1797? (can rename commits)

File: picasso/src/main/java/com/squareup/picasso3/Request.java [hunk 1]
```diff
@@ -480,6 +492,54 @@ public Builder transform(@NonNull List<? extends Transformation> transformations
...
+    Builder networkPolicy(@NonNull NetworkPolicy policy,
+        @NonNull NetworkPolicy... additional) {
+      if (policy == null) {
+        throw new IllegalArgumentException("Network policy cannot be null.");
+      }
+      this.networkPolicy |= policy.index;
+      if (additional == null) {
+        throw new IllegalArgumentException("Network policy cannot be null.");
+      }
+      if (additional.length > 0) {
```
Comment (swankjesse) [inline picasso/src/main/java/com/squareup/picasso3/Request.java]: nit: this if check is not necessary

File: picasso/src/main/java/com/squareup/picasso3/Request.java [hunk 2]
```diff
@@ -480,6 +492,54 @@ public Builder transform(@NonNull List<? extends Transformation> transformations
...
+    Builder memoryPolicy(@NonNull MemoryPolicy policy,
+        @NonNull MemoryPolicy... additional) {
+      if (policy == null) {
+        throw new IllegalArgumentException("Memory policy cannot be null.");
+      }
+      this.memoryPolicy |= policy.index;
+      if (additional == null) {
+        throw new IllegalArgumentException("Memory policy cannot be null.");
+      }
+      if (additional.length > 0) {
```
Comment (swankjesse) [inline picasso/src/main/java/com/squareup/picasso3/Request.java]: nit: this if check is not necessary

File: picasso/src/main/java/com/squareup/picasso3/Request.java [hunk 3]
```diff
@@ -480,6 +492,54 @@ public Builder transform(@NonNull List<? extends Transformation> transformations
...
     }
 
+    /**
+     * Specifies the {@link MemoryPolicy} to use for this request. You may specify additional policy
+     * options using the varargs parameter.
+     */
+    Builder memoryPolicy(@NonNull MemoryPolicy policy,
+        @NonNull MemoryPolicy... additional) {
+      if (policy == null) {
+        throw new IllegalArgumentException("Memory policy cannot be null.");
```
Comment (swankjesse) [inline picasso/src/main/java/com/squareup/picasso3/Request.java]: "policy == null" ?

File: picasso/src/main/java/com/squareup/picasso3/Request.java (hunk 2 above)
Comment (swankjesse) [reply picasso/src/main/java/com/squareup/picasso3/Request.java]: though this is not new code!

File: picasso/src/main/java/com/squareup/picasso3/Request.java (hunk 3 above)
Comment (JakeWharton) [reply picasso/src/main/java/com/squareup/picasso3/Request.java]: Also NPE not IAE

Comment (NightlyNexus) [reply picasso/src/main/java/com/squareup/picasso3/Request.java]: Yeah, I can fix these. Just existing code I pasted.

**Comment by jrodbx** [comment]
Closed in https://github.com/square/picasso/pull/1856.

**Comment by NightlyNexus** [comment]
nice, thanks!

**Pull request closed without merging**